pandas/_libs/tslibs/conversion.pyx

```diff
@@ -297,9 +297,18 @@ cdef _TSObject convert_to_tsobject(object ts, tzinfo tz, str unit,
         if ts == NPY_NAT:
             obj.value = NPY_NAT
         else:
-            ts = cast_from_unit(ts, unit)
-            obj.value = ts
-            pandas_datetime_to_datetimestruct(ts, NPY_FR_ns, &obj.dts)
+            in_reso = abbrev_to_npy_unit(unit or "ns")
+            out_reso = get_supported_reso(in_reso)
+            try:
+                value = convert_reso(ts, in_reso, out_reso, False)
```

Comment:
id probably just shove this into cast_from_unit

Reply:
in there it says

https://github.com/pandas-dev/pandas/blob/ef0eaa45a999d1ec0e1af0db66998e215f87a2dd/pandas/_libs/tslibs/np_datetime.pyx#L610-L614

though, are you OK with changing this?

Reply:
I don't understand the question

Reply:
OK, I'll try to clarify

You suggested moving
```
            try:
                value = convert_reso(ts, in_reso, reso, False)
            except OverflowError:
                # GH#26651 re-raise as OutOfBoundsDatetime
                raise OutOfBoundsDatetime(
                    f"cannot convert input {ts} with the unit '{unit}'"
                )
```
inside `cast_from_unit`

So, in `cast_from_unit`, I would do
```
try:
    res_value = value * mult 
except OverFlowError:
    raise OutOfBoundsDatetime(
        f"cannot convert input {ts} with the unit '{unit}'"
    )
```

I was going to do that, but then noticed the comment 
```
# Note: caller is responsible for re-raising as OutOfBoundsTimedelta 
```
which makes it look like this try-except should indeed be done outside `convert_reso`

Reply:
> should indeed be done outside convert_reso

Correct.  cast_from_unit _is_ outside convert_reso.

Reply:
not really sure how to put this inside `cast_from_unit`, it looks like it does something different

could you show what you had in mind please? (else I'll keep at it)

Reply:
ok, might have got, thanks for your patience - checking a few things, and updating

Reply:
right, might be working now - is this what you had in mind?